src/main/java/io/nats/client/SubscriptionImpl.java

```diff
@@ -51,6 +51,10 @@
     BlockingQueue<Message> mch;
     Condition pCond;
 
+    // Ideally, this should be in AsyncSubscriptionImpl, but because of
+    // locking and to reduce code duplication, leave this here.
+    private MsgDeliveryWorker dlvWorker;
+
```

Comment:
A `SubscriptionImpl` instance will never need a `MsgDeliveryWorker` unless it is also an `AsyncSubscription` - seems like implementation code not needed by synchronous subscription instances should be there.  Really, the shared code used by both should be factored into an internal class both of them *have* rather than *are*, and `AsyncSubscriptionImpl` should not subclass `SubscriptionImpl`.  This works, but it is only going to get messier and harder to maintain as years go by.

Reply:
That's the one I am not happy with. I originally put the worker only in the `AsyncSubscriptionImpl` class.

But I had then to check in connection's `processMsg()` what is the instance of the subscription in order to get the delivery worker (since then the getter would only be defined in the `AsyncSubscriptionImpl` class). 

Finally, I have assumed that I should never get worker's lock -> sub's lock sequence, but only the opposite, which would then force me to duplicate all `SubscriptionImpl` getters/setter that need to grab the worker's lock (to respect the assumed locking sequence). Looking more into this, I see no reason to do the other way around, which then means that I would have to overload them all but simply get the worker's lock (if worker is assigned, which is an immutable field) and then call `super` on the parent method.

Doing the review of the above, I realized that I missed protecting subscription `closed` flag with worker lock when appropriate.

Thanks for the review, I will update the PR.